### PR TITLE
🔀 Add `MergeModelCallBack`

### LINKED ISSUE #2241: Add model merging callback
### Feature request

Add a `MergeModelCallback` that merges the reference model with the current policy and optionally pushes the merged checkpoint to the Hub. This could be done on step/epoch end and/or the end of training. Implementation-wise, we could use Arcee's `mergekit` lib and include it as an optional dependency: https://github.com/arcee-ai/mergekit

### Motivation

Various papers show that model merging can non-trivially improve performance, especially if the models belong to the same architecture:

* https://arxiv.org/abs/2410.10801
* https://arxiv.org/abs/2406.16768 (for reward models)

### Your contribution

Open to the community!

### PR DESCRIPTION
**Pull request by August-murr**
# What does this PR do?
Fixes #2241

Since the focus was on replicating the checkpoint merging methods from [the paper](https://arxiv.org/abs/2410.10801), I have covered only Linear, TIES, SLERP, and DARE-TIES merging methods. These were the ones used in the paper and were primarily tested with the DPOTrainer.

Please provide feedback on any issues and suggest improvements regarding the structure of the files in this PR, as well as the documentation, to enhance clarity.

Here's an example of usage:

```bash
pip install trl[mergekit]
```
since it's an optional dependency

```python
from trl.mergekit_utils import MergeConfig
Trainer = DPOTrainer(...)
config = MergeConfig("ties")
config.target_model_path = "path_to_target_model" #if none is provided, the reference model will be used
config.policy_model_weight = 0.7 #optional
config.target_model_weight = 0.3 #optional 
merge_callback = MergeModelCallBack(config,push_to_hub=True,merge_at_every_checkpoint=True)
Trainer.add_callback(callback)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?
@lewtun 
@qgallouedec

### CODE REVIEW

File: setup.py
Comment (August-murr) [inline setup.py:106]: This is not ideal and very messy, but it was the only way I could install mergekit from source since "pip" installed an older version

**Comment by August-murr** [comment]
@lewtun 
@qgallouedec 
Feedback would be appreciated!

File: setup.py
Comment (lewtun) [reply setup.py:106]: The `mergekit` maintainer has kindly made a new release. Would you mind setting that with a lower bound and seeing if the callback still works?

Comment (August-murr) [reply setup.py]: It works the same overall, but `mergekit` 0.0.5 uses an older version of `accelerate`, which breaks training unless you upgrade `accelerate` afterward. installing `mergekit` from source doesn't cause this problem

Comment (cg123) [reply setup.py]: Sorry about that! 0.0.5.1 targets the current version of `accelerate`.

**Comment by qgallouedec** [comment]
Thanks a lot @August-murr for the work. Can you add documentation, and test?

File: trl/trainer/callbacks.py
Comment (qgallouedec) [inline trl/trainer/callbacks.py:433]:
```suggestion
class MergeModelCallback(TrainerCallback):
```

**Review by qgallouedec — CHANGES_REQUESTED** [review]

**Comment by August-murr** [comment]
> Thanks a lot @August-murr for the work. Can you add documentation, and test?

I've already added most of the docs, as for the tests, unfortunately I won't be able to do it for a few days and if nobody else added them, I'll do it later.

**Comment by August-murr** [comment]
The tests I added validate the success of the merge and I could expand it if necessary.
I also added docs to the callbacks file but was unable to produce the HTML file similar to the [callback docs](https://huggingface.co/docs/trl/main/en/callbacks) so I'd appreciate it if you could confirm whether the docs are properly generated or not.

File: tests/test_callbacks.py
Comment (lewtun) [inline tests/test_callbacks.py:231]: This dataset is a better choice to keep the tests running fast:

```suggestion
        self.dataset = load_dataset("trl-internal-testing/zen", "standard_preference", split="train")
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:11]: Can we reuse the `get_last_checkpoint()` [method](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/trainer_utils.py#L212) from `transformers` or is checking for the last modification time necessary here?

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:43]: nit: let's add the types
```suggestion
def upload_model_to_hf(folder_path : str, repo_id: str):
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:55]: Could you please add a docstring like we do for our model training configs: https://github.com/huggingface/trl/blob/623963126be5598bd1eea4ec82b43447fcc11535/trl/trainer/dpo_config.py#L34

This way it renders nicely in the docs :)

It would be good to provide some links to the various supported methods

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:86]:
```suggestion
    def create_merge_config_linear(self) -> MergeConfiguration:
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:90]: Technically this method has no args, so I'd remove this part of the docstring

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:97]: Can also be removed with the type hint above

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:115]:
```suggestion
    def create_merge_config_ties(self) -> MergeConfiguration:
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:120]: Same as above: we can remove the args

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:130]: Same as above: we can remove this once we include the typehint

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:190]:
```suggestion
    def create_merge_config_dare_ties(self) -> MergeConfiguration:
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:195]: Remove

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:205]: Remove

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:269]: Remove

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:265]:
```suggestion
    def create_merge_config_slerp(self) -> MergeConfiguration:
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:275]: Remove

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:320]:
```suggestion
    def create(self) -> MergeConfiguration:
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:84]: turbo nit:
```suggestion
            raise ValueError(f"Unsupported merge method: {method}")
```

File: trl/mergekit_utils.py
Comment (lewtun) [inline trl/mergekit_utils.py:330]: nit: we use snakecase for methods:

```suggestion
def merge_models(config : MergeConfig, out_path : str):
```

File: trl/trainer/callbacks.py
Comment (lewtun) [inline trl/trainer/callbacks.py:465]: nit: we prefer f-strings in trl

```suggestion
            output_path = f"{policy_model_path}/merged"
```

File: trl/trainer/callbacks.py
Comment (lewtun) [inline trl/trainer/callbacks.py:482]:
```suggestion
            output_path = f"{policy_model_path}/merged"
```

**Review by lewtun — CHANGES_REQUESTED** [review]
Thanks for adding this callback @August-murr - the implementation is very clean! I've left some minor comments on the PR and then I think it's good to merge (pun intended :))

Can you also fix the tests - it seems to be an issue with the path creation on Windows. Also can you add the callback to the docs here: https://github.com/huggingface/trl/blob/main/docs/source/callbacks.mdx

File: trl/__init__.py
Comment (qgallouedec) [inline trl/__init__.py:91]: Please keep the alphabetical order

File: trl/__init__.py
Comment (qgallouedec) [inline trl/__init__.py:95]:
```suggestion
    "trainer.callbacks": ["MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback"],
```

File: trl/__init__.py
Comment (qgallouedec) [inline trl/__init__.py:185]:
```suggestion
    from .trainer.callbacks import MergeModelCallback, RichProgressCallback, SyncRefModelCallback
```

File: trl/__init__.py
Comment (qgallouedec) [inline trl/__init__.py:181]: Please keep the alphabetical order

File: trl/mergekit_utils.py
Comment (qgallouedec) [reply trl/mergekit_utils.py:1]: or just run `make precommit`, it will make it for you

File: trl/trainer/__init__.py
Comment (qgallouedec) [inline trl/trainer/__init__.py:27]:
```suggestion
    "callbacks": ["LogCompletionsCallback", "MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback", "WinRateCallback"],
```

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:56]: Can we have a default value here?

File: trl/mergekit_utils.py
Comment (qgallouedec) [reply trl/mergekit_utils.py:55]: Useful ressource: https://github.com/huggingface/trl/pull/1944#issue-2471984781

File: trl/mergekit_utils.py
Comment (August-murr) [reply trl/mergekit_utils.py]: just replaced it

File: trl/mergekit_utils.py
Comment (August-murr) [reply trl/mergekit_utils.py:22]: I got GPT to add the docstring. 
is it ok?

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:1]:
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:44]:
```suggestion

            - `"linear"`: Linearly combines two models with specified weights.
```

otherwise it doesn't render properly

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:48]:
```suggestion

    Note:
```

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:43]:
```suggestion
            Merge method to use. Supported methods include:
```

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:64]: you can drop all of these

File: trl/mergekit_utils.py
Comment (qgallouedec) [reply trl/mergekit_utils.py:34]: Sure!

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:58]:
```suggestion
        policy_model_density (`List[float]`): Density parameters for the policy model (for `ties` and `dare_ties`).
```

They are not optional per se. See typing doc (same for the other)

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:90]: Why do you need this?

File: trl/mergekit_utils.py
Comment (qgallouedec) [inline trl/mergekit_utils.py:150]: Can we omit the values that are none?

File: trl/trainer/callbacks.py
Comment (qgallouedec) [inline trl/trainer/callbacks.py:459]: push_to_hub should be gotten from the training args  I think

File: trl/mergekit_utils.py
Comment (August-murr) [reply trl/mergekit_utils.py:150]: I left the values as `None` so we can add support for LoRA or other MergeKit features in the future if needed.

File: trl/trainer/callbacks.py
Comment (August-murr) [reply trl/trainer/callbacks.py:459]: push_to_hub from training args pushes the trained model 
this one pushes the merged model

File: trl/mergekit_utils.py
Comment (August-murr) [reply trl/mergekit_utils.py:90]: default values taken from examples in the Mergekits repo.
can be set or changed by the user

**Review by lewtun — APPROVED** [review]
Thanks for iterating @August-murr ! The PR LGTM now and once the CI is green & @qgallouedec approves, I think we can merge it

**Comment by August-murr** [comment]
> Thanks for iterating @August-murr ! The PR LGTM now and once the CI is green & @qgallouedec approves, I think we can merge it

The tests without optional dependency failed since Mergekit is an optional dependency

**Comment by kashif** [comment]
@August-murr in the `import_utils` you can define a new `is_mergekit_available` helper and then in the tests you can skip the tests if its not available

**Comment by qgallouedec** [comment]
Like here:

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/trl/import_utils.py#L39-L40

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/testing_utils.py#L42-L46

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/test_judges.py#L62-L63

Don't hesitate to ask for help if you want the maintainers to do it for you.

**Comment by August-murr** [comment]
> Like here:
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/trl/import_utils.py#L39-L40
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/testing_utils.py#L42-L46
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/test_judges.py#L62-L63
> 
> Don't hesitate to ask for help if you want the maintainers to do it for you.

Done!

File: tests/testing_utils.py
Comment (qgallouedec) [inline tests/testing_utils.py:54]:
```suggestion
    Decorator marking a test that requires Mergekit. Skips the test if mergekit is not available.
    """
    return unittest.skipUnless(is_mergekit_available(), "test requires mergekit")(test_case)
```

File: trl/__init__.py
Comment (qgallouedec) [inline trl/__init__.py:95]:
```suggestion
    "trainer.callbacks": ["MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback"],
```

**Comment by qgallouedec** [comment]
Nice, thanks! Just running some tests, waiting for the CI to be green, and we're good to merge (expect some commits from me on this branch)

File: trl/mergekit_utils.py
Comment (qgallouedec) [reply trl/mergekit_utils.py:76]: the default from the class definition is `None` (see https://github.com/arcee-ai/mergekit/blob/57e7d14e2a732f532970e2c9dada00e2d8f15a7a/mergekit/config.py#L91) so I would keep it like this

**Comment by qgallouedec** [comment]
Another question that raised while reviewing: while having a new configuration class while we can directly use the one from `mergekit`. I'm afraid that I could confuse the user, tempted to use:

```python
from mergekit import MergeConfiguration
from trl import MergeModelCallback
    
merge_callback = MergeModelCallback(MergeConfiguration())
```

**Comment by August-murr** [comment]
> Another question that came up during the review: why have a new configuration class when we can use the mergekit one directly? I'm afraid of confusing the user, tempted to use :
> 
> ```python
> from mergekit import MergeConfiguration
> from trl import MergeModelCallback
>     
> merge_callback = MergeModelCallback(MergeConfiguration())
> ```

Actually, ease of use for the user was the reason why I had to write the class in mergekit_utils since mergekit uses a yaml file to get it's Merge config, which is easier to implement but more complicated for the user.

and if you wanted to use `MergeConfiguration` directly from mergekit:
```python
from mergekit.config import MergeConfiguration

merge_config_dict = {
    "dtype": "float16",
    "merge_method": "linear",
    "models": [
        {"model": "path_to_model_1", "parameters": {"weight": 0.4}},
        {"model": "path_to_model_2", "parameters": {"weight": 0.6}},
    ],
}

config = MergeConfiguration.model_validate(merge_config_dict)
```
As you add more parameters to the configuration, the dictionary becomes increasingly nested.

The current implementation, although harder to maintain, simplifies everything for the user:
```python
from trl.mergekit_utils import MergeConfig
config = MergeConfig("linear")
config.policy_model_weight = 0.4
config.target_model_weight = 0.6
```

**Comment by qgallouedec** [comment]
That makes sense.
Do you think we can get the best of both worlds by making `trl.MergeConfig` inherits from `mergekit.config.MergeConfigurationMergeConfig`?

**Comment by August-murr** [comment]
> That makes sense.
> Do you think we can get the best of both worlds by making `trl.MergeConfig` inherits from `mergekit.config.MergeConfigurationMergeConfig`?

I'll figure it out.

**Comment by qgallouedec** [comment]
LGTM thanks!
I've just applied some minor refinements:

- compat with windows file path
- sort imports and function
- common method for saving and pushing in the callback
- add "trl" to model tags

**Comment by August-murr** [comment]
@qgallouedec 
About the failed tests:
The tests do not fail on Ubuntu; they only fail on Windows. I realized that the issue arose from a permission error from the temporary directory when trying to delete the merged files, specifically the `model.safetensors.`

**Comment by qgallouedec** [comment]
> @qgallouedec About the failed tests: The tests do not fail on Ubuntu; they only fail on Windows. I realized that the issue arose from a permission error from the temporary directory when trying to delete the merged files, specifically the `model.safetensors.`

Ah thanks, I was debugging, but I don't have access to windows vm right now (explains https://github.com/huggingface/trl/pull/2282/commits/fa5bafe617793ed340303cf0ebded6ac03cab39f). Any idea how to solve it?

**Comment by qgallouedec** [comment]
Found a solution with a57d88a1b317785fa85e3b09bd463ecb0b9eef06

**Comment by August-murr** [comment]
@qgallouedec 
Sorry I wasn't able to sort it out myself.

**Pull request merged**
